### ISSUE TITLE
Does `SighashCache::*_encode_signing_data_to` need `async` support?

### ISSUE THREAD

**Issue by Kixunil** [rust-bitcoin/rust-bitcoin#3106]
I think these are realistically only used to compute hashes and maybe debug the preimage of the hashes both of which work just fine without `async`. But I might be missing something.

I was thinking that instead of doing encoders or using `io` traits just accept a closure `FnMut(&[u8]) -> Result<(), E>` with arbitrary `E` (so infallible users can use `!`). This would be much simpler and avoid adding new traits or changing the code too much.

**Comment by apoelstra** [comment]
Yeah, that sounds good to me. I'm not sure what the implications for `async` would be of this.

**Comment by Kixunil** [comment]
Basically for `async` we'd have to write `Encoder` impl which I'd rather avoid for something that complex without generators/coroutines or write the same code twice which I'd absolutely hate (the number of times breaking single source of truth kicked me or someone else in the butt is crazy). And I very highly doubt there's a reason to do so and if there is it's probably not on embedded hardware so serializing to `Vec` and then sending that is not really that big deal, especially during signing which already takes some time. At least as a stop gap. But I suspect Rust will stabilize coroutines before this becomes a thing. :D

**Comment by apoelstra** [comment]
Yeah, let's not do any extra work (especially duplicated work!) to support `async` here. I believe you're correct that there's no need for it.
